libgalois/src/analytics/cdlp/cdlp.cpp

```diff
@@ -0,0 +1,402 @@
+/*
+ * This file belongs to the Galois project, a C++ library for exploiting
+ * parallelism. The code is being released under the terms of the 3-Clause BSD
+ * License (a copy is located in LICENSE.txt at the top-level directory).
+ *
+ * Copyright (C) 2018, The University of Texas at Austin. All rights reserved.
+ * UNIVERSITY EXPRESSLY DISCLAIMS ANY AND ALL WARRANTIES CONCERNING THIS
+ * SOFTWARE AND DOCUMENTATION, INCLUDING ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR ANY PARTICULAR PURPOSE, NON-INFRINGEMENT AND WARRANTIES OF
+ * PERFORMANCE, AND ANY WARRANTY THAT MIGHT OTHERWISE ARISE FROM COURSE OF
+ * DEALING OR USAGE OF TRADE.  NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH
+ * RESPECT TO THE USE OF THE SOFTWARE OR DOCUMENTATION. Under no circumstances
+ * shall University be liable for incidental, special, indirect, direct or
+ * consequential damages or loss of profits, interruption of business, or
+ * related expenses which may arise from use of Software or Documentation,
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#include "katana/analytics/cdlp/cdlp.h"
+
+#include <boost/unordered_map.hpp>
+
+#include "katana/ArrowRandomAccessBuilder.h"
+#include "katana/TypedPropertyGraph.h"
+
+using namespace katana::analytics;
+
+namespace {
+const unsigned int kInfinity = std::numeric_limits<unsigned int>::max();
+
+/// Limited number of iterations to limit the oscillation of the label
+/// in Synchronous algorithm. We dont need to limit it in Asynchronous algorithm.
+/// Set to 10 same as Graphalytics benchmark.
+const unsigned int kMaxIterations = 10;
+
+struct CdlpSynchronousAlgo {
+  using CommunityType = uint64_t;
+  struct NodeCommunity : public katana::PODProperty<CommunityType> {};
+
+  using NodeData = std::tuple<NodeCommunity>;
+  using EdgeData = std::tuple<>;
+  typedef katana::TypedPropertyGraph<NodeData, EdgeData> Graph;
+  typedef typename Graph::Node GNode;
+  using BiDirGraphView = katana::TypedPropertyGraphView<
```

Comment:
Please use Undirected view and then you'd write only one loop for both in and out edges.

Reply:
It is merged! perfect, will do.